Refactor gazel to write all BUILD files in one pass at the end, and add --validate mode

ixdy opened the pull request:
Part of working towards https://github.com/mikedanese/gazel/issues/12.

This PR should functionally be mostly a no-op (though we don't have any tests to confirm...). I've manually tested on kubernetes/test-infra, where it seems to DTRT.

The only functional change here is that `BUILD.bazel` files are supported, if they already exist. When creating new files, we still default to `BUILD`.

ixdy committed:
Refactor gazel to write all BUILD files in one pass at the end

mikedanese commented on gazel/gazel.go:
```diff
@@ -37,11 +37,14 @@ func main() {
+		glog.Fatalf("unable to build vendorer: %v", err)
 	}
 
 	if len(flag.Args()) == 1 {
```
I'm not sure if this feature is useful anymore now that gazel is much faster.

ixdy replied:
removed.

mikedanese commented on gazel/gazel.go:
```diff
@@ -429,6 +425,16 @@ func (v *Vendorer) extractDeps(deps []string) *bzl.ListExpr {
 }
 
+func (v *Vendorer) reconcileAllRules() error {
+	for path, rules := range v.newRules {
```
Can we write these rules in a deterministic order?

ixdy replied:
now sorted!

mikedanese commented:
> deterministic order

this is the only comment I care to address. Otherwise LGTM

spxtr commented:
Looks good. Does this make it easier to add a "--validate" flag?

ixdy committed:
A few additional no-op cleanups

ixdy committed:
review feedback: remove single-arg case, sort package paths before writing

ixdy commented:
I added a few more no-oppy changes

ixdy commented:
I added a few more no-oppy changes in 0a3220f3aec92bccbabe593ac1b041c93a96acc4, basically things I realized I needed as I kept working:
* clean paths added to `v.newRules` to avoid accidental duplicates from different codepaths
* rename `buildName` to `findBuildFile` and make it return whether the file exists
* move `"automanaged"` into a constant

And I addressed your feedback in the last comment. If you think this still LG, I (or you) can squash.

@spxtr what would `--validate` do?

spxtr commented:
`--validate` exits with a non-zero exit code if it would write any changes.

ixdy commented:
ah, this PR makes it both easier and harder:
* easier in that `reconcileAllRules` can determine if any changes were needed
* harder in that `writeFile` no longer returns whether it wrote anything, cause I refactored the printing bit into it.

ixdy committed:
make writeFile, ReconcileRules, etc return bools again

ixdy commented:
ok, `reconcileAllRules` now returns whether any files were written, but I'm ignoring it. will probably do `--validate` in a separate PR.

ixdy committed:
Add --validate mode which exits nonzero if any files need updating.

--validate implies --dry-run

spxtr commented on gazel/gazel.go:
```diff
@@ -49,9 +53,17 @@ func main() {
+			fmt.Fprintf(os.Stderr, "\n%d BUILD files not up-to-date.\n", written)
+			os.Exit(1)
+		}
+		fmt.Fprintf(os.Stderr, "All BUILD files up-to-date.\n")
```
nit: print nothing on success.

ixdy replied:
I kinda like having the affirmation that everything is OK, but I can remove this.

spxtr replied:
I don't have a source, but I'm under the impression that if a validation-type-thing succeeds, the Unix Way is to exit 0 and print nothing. gofmt and go vet follow this, for instance.

ixdy replied:
fair enough. this has been removed.

ixdy commented:
ok, I lied, and I added `--validate` now too.

```console
~/.../src/k8s.io/kubernetes $ gazel --validate ; echo exit $?
DRY-RUN: wrote "cmd/clicheck/BUILD"
DRY-RUN: wrote "cmd/cloud-controller-manager/BUILD"
DRY-RUN: wrote "cmd/gendocs/BUILD"
...
DRY-RUN: wrote "third_party/forked/golang/expansion/BUILD"
DRY-RUN: wrote "third_party/forked/golang/reflect/BUILD"
DRY-RUN: wrote "vendor/BUILD"

512 BUILD files not up-to-date.
exit 1
~/.../src/k8s.io/test-infra $ gazel --validate ; echo exit $?
All BUILD files up-to-date.
exit 0
```

ixdy committed:
--validate: silent on success

ixdy committed:
Fix diff order

mikedanese commented:
I read and monkey tested this and seems to be a noop. LGTM
